### ISSUE TITLE
Investigating Multilingual Post-Training Data Performance

### ISSUE THREAD

**Issue by ferreirafabio** [OpenEuroLLM/Taskboard#186]
## Overview

We're investigating how to add multilingual capabilities to OLMo-3-7B without catastrophic forgetting of English. Starting from our reproduced instruct SFT checkpoint, we continue training on mixtures of English + EU-language data at varying ratios.

### Background: Dolci-Instruct-SFT is 93% English

The base checkpoint's training data (2.15M samples) is overwhelmingly English with a long tail of ~160 languages, each below ~0.4% (Spanish is 2nd largest with ~0.39% (instruct) ~0.41% (think)).

Count-based:
![Dolci-Instruct-SFT language counts](https://github.com/ferreirafabio/open-instruct/blob/main/oellm/experiments/dolci_distribution/results/dolci_language_counts.png?raw=true)

Percentage-based (w.r.t. the respective dataset):
![Dolci-Instruct-SFT language percentage](https://github.com/ferreirafabio/open-instruct/blob/main/oellm/experiments/dolci_distribution/results/dolci_language_distribution.png?raw=true)

### Setup

| | |
|---|---|
| **Base checkpoint** | OLMo-3-7B-Instruct-SFT (reproduced, [details](https://github.com/allenai/open-instruct/issues/1352#issuecomment-2823953936)) |
| **Judge (Elo)** | Qwen3.5-27B (dense, via vllm serve, both orderings, 8k max tokens) |
| **Judge (winrate)** | Qwen3-30B-A3B-Instruct-2507 (VLLM, both orderings, 8k max tokens) |
| **Elo benchmark** | LMArena, Bradley-Terry, 100 bootstraps, 200 battles/language, 12 EU languages |
| **Winrate benchmarks** | m-arena-hard-EU (6000 prompts, 12 EU languages), arena-hard (500 English prompts) |
| **Training languages** | en + de, es, fr, it, pt, pl, nl, cs (ratio varies by experiment) |
| **Eval languages** | en, de, es, fr, it, pt, pl, nl, cs, ro, el, uk (12 languages available in LMArena) |
| **Held-out languages** | ro, el (zero-shot transfer test, not in training data) |

### Experiment tracks

| Track | Question | Design |
|---|---|---|
| **A** | Preliminary: does the English/EU ratio matter? | 90/80/70% English, ~95k samples, fusion-synth only, EU share split equally across 8 languages |
| **B** | Does more diverse data help? | Same ratios, ~490k samples, fusion-synth + wildchat, lmsys-chat, oasst2 |
| **C** | Is English regression caused by EU data, or by continued SFT itself? | 100% English control (no EU data), same total samples |
| **D** | Does replaying the base checkpoint's English data reduce forgetting? | Same ratios as Track A, but English from Dolci-Instruct-SFT (base checkpoint's own training data) |
| **E** | Does Dolci replay scale with more data? | Same as Track D, but ~490k samples instead of ~95k |

### Full experiment matrix

| Exp. | En/EU | N | Elo | Elo en | Elo w/o en | EU WR% | en WR% |
|---|---|---|---|---|---|---|---|
| **Baseline** | — | — | 741±9 | **950±21** | 722±10 | 50% | 50% |
| **A1-90en** | 90/10 | 94.7k | 702±10 | 771±32 | 692±10 | **54.8%** | 14.1% |
| **A2-80en** | 80/20 | 93.6k | 704±11 | 769±30 | 703±11 | **57.2%** | 12.4% |
| **A3-70en** | 70/30 | 91.7k | 713±10 | 766±29 | 689±11 | **58.8%** | 13.3% |
| **B1-90en** | 90/10 | 491k | 720±9 | 789±26 | 708±10 | 53.6% | 13.0% |
| **B2-80en** | 80/20 | 473k | 722±9 | 797±26 | 722±10 | 52.9% | 14.2% |
| **C0-100en** | 100/0 | 94.7k | 670±11 | 791±29 | 681±11 | 48.9% | 11.8% |
| **D1-90en** | 90/10 | 94.7k | **751±8** | **942±20** | 716±11 | **63.4%** | **54.6%** |
| **D2-80en** | 80/20 | 93.6k | **751±8** | **956±20** | **725±12** | **62.0%** | **54.6%** |
| **D3-70en** | 70/30 | 91.7k | **753±9** | **963±21** | **731±11** | **63.5%** | **54.3%** |
| **E1-90en** | 90/10 | 491k | **758±9** | **965±21** | **740±9** | **59.9%** | **57.0%** |
| **E2-80en** | 80/20 | 474k | **759±8** | **931±24** | **725±9** | **56.6%** | **58.2%** |
| **E3-70en** | 70/30 | 455k | **751±9** | **940±22** | **726±9** | 53.1% | **58.6%** |

Elo: Qwen3.5-27B judge. "en" = English-only (200 battles). "w/o en" = 11 non-English languages.

EU WR% / en WR% = winrate vs baseline (Qwen3-30B-A3B judge). 50% = parity.

### Findings

1. **Dolci replay preserves English** (Track D): English Elo stays at 942-963 (baseline is 950), English arena-hard winrate ~55%. Without replay (Track A), English drops to 766-771 Elo / ~13% winrate. The source of the English data matters, the En/EU ratio doesn't.
2. **The D-track advantage is almost entirely English**: per-language Elo shows A vs D gap is ~190 points on English. On trained EU languages, they're within CIs.
3. **Scaling helps modestly**: E1 (90/10 at 491k) reaches the highest overall Elo (758). More EU data at scale (E2/E3) shows diminishing returns.
4. **More diverse data doesn't help** (Track B): 5× more data from wildchat/lmsys/oasst2 → similar Elo to Track A.
5. **Continued SFT itself causes forgetting** (Track C): C0 (100% English, no EU data at all) still drops English Elo to 791. The regression is not caused by EU data but by training on non-Dolci English.
6. **Transfer**: Romanian (held-out, Latin script) transfers well (Elo 700-890). Greek (held-out, different script) does not (540-690).

### Code

- Training configs: [`oellm/configs/`](https://github.com/ferreirafabio/open-instruct/tree/main/oellm/configs)
- Evaluation scripts: [`oellm/evaluations/benchmarks/`](https://github.com/ferreirafabio/open-instruct/tree/main/oellm/evaluations/benchmarks)
- Full results: [`multilingual_eu_results.md`](https://github.com/ferreirafabio/open-instruct/blob/main/oellm/experiments/multilingual_eu/results/multilingual_eu_results.md)

**Comment by ferreirafabio** [comment]
## Update: Track F with extreme En/EU ratios

Following up on the Track A–E results above, we mapped the full English/EU ratio curve (100% → 0% English) with Dolci replay.

### Experiment tracks (recap + new)

| Track | Question | Design |
|---|---|---|
| **A** | Preliminary: does the English/EU ratio matter? | 90/80/70% English, ~95k samples, fusion-synth only, EU share split equally across 8 languages |
| **B** | Does more diverse data help? | Same ratios, ~490k samples, fusion-synth + wildchat, lmsys-chat, oasst2 |
| **C** | Is English regression caused by EU data, or by continued SFT itself? | 100% English control (no EU data), same total samples |
| **D** | Does replaying the base checkpoint's English data reduce forgetting? | Same ratios as Track A, but English from Dolci-Instruct-SFT (base checkpoint's own training data) |
| **E** | Does Dolci replay scale with more data? | Same as Track D, but ~490k samples instead of ~95k |
| **F** | How do extreme En/EU ratios affect performance? | 100/75/50/25/0% English, ~500k target, 6 EU langs (de,es,fr,it,pt,pl), Dolci replay |

Note: F track sample counts are not fully matched because Polish (7k available) and Italian (29k) cap the high EU ratio configs.

### Results

| Exp. | En/EU | N | Elo | Elo en | Elo w/o en | EU WR% | en WR% |
|---|---|---|---|---|---|---|---|
| **F1-100en** | 100/0 | 500k | 728±9 | **954±22** | 695±11 | 52.2% | **57.6%** |
| **F2-75en** | 75/25 | 486k | **747±10** | **947±22** | 711±10 | 51.4% | **59.7%** |
| **F3-50en** | 50/50 | 453k | 726±9 | **935±21** | 700±12 | 44.7% | **55.9%** |
| **F4-25en** | 25/75 | 359k | 739±9 | **912±22** | 685±11 | 43.3% | **57.2%** |
| **F5-0en** | 0/100 | 234k | 677±11 | 762±31 | 680±11 | 36.6% | 9.7% |

Elo: Qwen3.5-27B judge, LMArena Bradley-Terry, 200 battles/lang. EU/en WR%: Qwen3-30B-A3B judge, vs baseline.

### Per-language Elo (all tracks)

![Per-language heatmap](https://github.com/ferreirafabio/open-instruct/blob/main/oellm/experiments/multilingual_eu/results/plots/fg_per_language_heatmap.png?raw=true)

### Elo vs English/EU ratio (F-Track)

![Ratio curve](https://github.com/ferreirafabio/open-instruct/blob/main/oellm/experiments/multilingual_eu/results/plots/fg_ratio_curve.png?raw=true)

### English vs EU tradeoff (all tracks)

![Tradeoff scatter](https://github.com/ferreirafabio/open-instruct/blob/main/oellm/experiments/multilingual_eu/results/plots/fg_tradeoff_scatter.png?raw=true)

### Observations

1. **English Elo scales with English ratio**: F1 (954) > F2 (947) > F3 (935) > F4 (912) > F5 (762). Dolci replay preserves English well up to ~50% EU, then degrades. At 0% English, collapses to ~760.
2. **75/25 is the overall sweet spot**: F2 achieves the highest overall Elo (747) and competitive EU WR% (51.4%).
3. **Pure EU training hurts EU performance too**: F5 (0% English) EU WR = 36.6%, below baseline (50%). Without English, the model loses general instruction following ability.
4. **Non English Elo is relatively flat across ratios**. The main lever is English preservation. EU languages don't clearly benefit from more EU training data in the mix.

Full results with all tracks: [multilingual_eu_results.md](https://github.com/ferreirafabio/open-instruct/blob/main/oellm/experiments/multilingual_eu/results/multilingual_eu_results.md#full-experiment-matrix)

**Comment by abhash-er** [comment]
## Update Track G Experiments


The next set of experiments, that we call **Track G** continually-trains `allenai/Olmo-3-7B-Instruct-SFT`. We generate synthetic instruction following dataset available at [openeurollm/EU-Instruct-Synthetic](https://huggingface.co/datasets/openeurollm/EU-Instruct-Synthetic) with constitution of 11 EU languages with each language representing around ~150k examples. (#345 )

| Code | Language | Examples |
| :--- | :--- | ---: |
| cs | Czech | 150,129 |
| de | German | 135,786 |
| el | Greek | 138,048 |
| es | Spanish | 132,736 |
| fr | French | 119,497 |
| it | Italian | 136,407 |
| nl | Dutch | 131,515 |
| pl | Polish | 146,247 |
| pt | Portuguese | 131,707 |
| ro | Romanian | 129,112 |
| uk | Ukrainian | 146,092 |
| **Total** | | **1,497,276** |****

> Note that each instruction is restricted to 4k sequence length

Out of this dataset, we create Dolci-Synthetic-Replay mixture with 500k and 1 Million instructions where we sample english data from the original [Dolci-Instruct-SFT](https://huggingface.co/datasets/allenai/Dolci-Instruct-SFT) in varying ratio from 0/25/50/75/100 and the rest of the samples are per-lang-equally sampled from the  [openeurollm/EU-Instruct-Synthetic](https://huggingface.co/datasets/openeurollm/EU-Instruct-Synthetic) dataset. 

Here's the overall winrate and ELO for LMArena-140k (2204 battles):

| Model | Scale | Overall Elo | EN Elo (independent) | W/O EN Elo | EN WR | EU WR | 
|---|---|---|---|---|---|---|
| baseline | — | 723±12 | 956±46 | 679±22 | 50 | 50 | 
| G1 100en | 500k | 723±12 | 961±46 | 677±23 | 49.2 | 47.0 |
| G1 100en | 1M | 725±12 | 950±43 | 685±19 | 51.2 | 50.4 | 
| G2 75en | 500k | 752±12 | 970±46 | 712±22 | 51.3 | 73.7 |
| G2 75en | 1M | 755±12 | 951±42 | 716±26 | 51.9 | 70.6 | 
| G3 50en | 500k | 757±12 | 946±42 | 723±18 | 47.4 | 76.0 |
| G3 50en | 1M | 767±12 | 975±46 | 724±23 | 49.5 | 73.7 | 
| G4 25en | 500k | 757±12 | 964±45 | 716±23 | 47.3 | 79.6 |
| G4 25en | 1M | 766±12 | 947±42 | 702±47 | 46.5 | 76.2 | 
| G5 0en | 500k | 750±12 | 925±42 | 713±23 | 44.1 | 75.8 |
| G5 0en | 1M | 759±12 | 915±41 | 722±21 | 44.1 | 77.1 | 


_Note: We use the same setting as above but with one difference- we measure ELO on 12 EU langauges (`en`, `de`, `es`, `fr`, `it`, `pt`, `pl`, `nl`, `cs`, `ro`, `el`, `uk`) rather than the 8 used in the previous step (`en`, `cs`, `de`, `es`, `fi`, `fr`, `it`, `sv`) and the English ELO is calculated independent from the overall (but right now is fully anchored) and the other ELOs are full-anchored ones._


Trends still hold clean:
- **EN WR** drops as we cut english: 51 -> 44 (expected regression)
- **EU WR** climbs as we add multilingual: 50 -> 77, peaks G4-25en-500k at 79.6
- overall ELO peaks with SOME multilingual (G3-50en-1M 767), and the english is weakest at G5-0en

Here is the result on ComparIA (92% french arena) with Qwen 3-30B as Judge (1171 battles):

 | Model | Scale | ComparIA Elo  |
|---|---|---|
| baseline | — | 730±1 |
| G1 100en | 500k | 717±1 |
| G1 100en | 1M | 719±1 |
| G2 75en | 500k | 742±1 |
| G2 75en | 1M | 741±1 |
| G3 50en | 500k | 746±1 |
| G3 50en | 1M | 745±1 |
| G4 25en | 500k | 752±1 |
| G4 25en | 1M | 753±1 |
| G5 0en | 500k | 767±1 |
| G5 0en | 1M | **772±1** |

ComparIA Elo rises MONOTONICALLY as english drops (`baseline` 730 ->`G5 0en` 772), opposite of the LMArena overall curve (which peaks mid at G3). But this makes sense as it is french-heavy arena rewards more multilingual training.

<details>
<summary> Training Details (Click to expand) </summary>
For training the models, we switch the stack from Nvidia H200s (Horeka) to AMD MI250x (LUMI). Find the code used to train <a  href="https://github.com/abhash-er/open-instruct/tree/lumi-trackg-tokenization-pipeline">here</a>.

- **base model**: OLMo-3-7B (d_model 4096, 32 layers, 32 heads, ffn 11008, reordered-norm block,
  qk-norm rms, YaRN rope theta=500000 factor=8, sliding-window pattern [4096,4096,4096,-1]), continued
  SFT on top of the reproduced Instruct-SFT.
- **optimizer**: AdamW, lr **8e-5**, betas (0.9, 0.95), eps 1e-8, weight_decay 0.0, max_grad_norm 1.0
- **schedule**: warmup 3% of steps, cosine decay to 0 (alpha_f=0), 2 epochs
- **batch / seq**: global batch **1,048,576 tokens** (1M), rank-microbatch 32768, seq_len **32768**
- **parallelism**: HSDP, shard_degree 4, param bf16 / reduce fp32, model compiled
- **seeds**: init_seed 33333, data-loader seed 34521
- **steps** (final, per run) -1M ≈ 2× the 500k because 2× the data: G1 696/1390, G2 680/1362, G3 668/1334, G4 652/1304, G5 636/1272 (500k / 1M)
- **Epochs** - 2
</details>

## Plots
### 1.  Elo vs English/EU Ratio

<img width="4458" height="1418" alt="Image" src="https://github.com/user-attachments/assets/b5f51acc-b5f8-4c06-8791-927b7f5a54e2" />

### 2. English vs EU Winrates

<img width="2209" height="1461" alt="Image" src="https://github.com/user-attachments/assets/743d1037-11d5-43f5-86d2-0a2dcbaba8ea" />

### 3. Per-Language (Independent) Elo heat map

<img width="3086" height="1400" alt="Image" src="https://github.com/user-attachments/assets/1706912b-e43d-436e-afad-f9066452e1a7" />

### 4. English vs EU tradeoff

<img width="2358" height="2061" alt="Image" src="https://github.com/user-attachments/assets/580adb32-1ce5-4b32-a6cc-d71ab8635d8c" />

## what's still open
- Q3-lmarena reruns: test job in the queue, waiting on it. resubmit the rest if it clears.
- To add the models - [OLMo-3-7B-Dolci-Translated-A-25EN](https://huggingface.co/openeurollm/OLMo-3-7B-Dolci-Translated-A-25EN) and [openeurollm/OLMo-3-7B-Dolci-Translated-A-75EN](https://huggingface.co/openeurollm/OLMo-3-7B-Dolci-Translated-A-75EN)

**Comment by abhash-er** [comment]
I also ran the models above with different judges and included the [OLMo-3-7B-Dolci-Translated-A-25EN](https://huggingface.co/openeurollm/OLMo-3-7B-Dolci-Translated-A-25EN) and [openeurollm/OLMo-3-7B-Dolci-Translated-A-75EN](https://huggingface.co/openeurollm/OLMo-3-7B-Dolci-Translated-A-75EN) models (Track F's best) with the Track G!

**Per-language Heatmap:**
| Qwen3.5-27B | Gemma-4-31B | Gemma-4-26B-A4B |
|---|---|---|
<img width="3126" height="2300" alt="Image" src="https://github.com/user-attachments/assets/399a2062-ade8-495c-93df-9386f2c9885c" />| <img width="3126" height="2300" alt="Image" src="https://github.com/user-attachments/assets/e822bb4e-ce4d-457a-ab45-70ec2f57faea" /> | <img width="3126" height="2300" alt="Image" src="https://github.com/user-attachments/assets/6fcffc19-5f64-48f8-ad18-43f318027dc4" />|

**EU vs Eng Tradeoff:**
| Qwen3.5-27B | Gemma-4-31B | Gemma-4-26B-A4B |
|---|---|---|
| <img width="2355" height="2061" alt="Image" src="https://github.com/user-attachments/assets/8126df5f-d10b-4662-94d7-750d82974e0d" /> | <img width="2355" height="2061" alt="Image" src="https://github.com/user-attachments/assets/97e0da89-6853-43e1-9380-83d4a7600c00" /> | <img width="2355" height="2061" alt="Image" src="https://github.com/user-attachments/assets/46add475-08d1-4209-9b97-4492318f3966" /> |
